internal/RF_Base.ecl

```diff
@@ -21,36 +22,134 @@ TreeNodeDat := Types.TreeNodeDat;
+  * while the derived modules (e.g. ClassificationForest and RegressionForest) overlay
+  * virtual attributes that perform work specific to the type of forest.
+  * The basic steps in building the tree are:
+  * - Organize the input data:
+  *   - Identify the work-items and number of training records and features
+  *   - Ensure that the record ids are sequential
+  *   - Create a map of each of the feature numbers for each work-item
+  * - Generate random bootstrap samples for each tree
+  * - Create numTree tree root nodes
+  * - Replicate the training data bootstraps to each tree root
+  * - Build the trees, one layer at a time by:
+  *   - Randomly choose the featuresPerNode features to evaluate for a given level
+  *   - Evaluate all possible splits on those features to find the feature and split-point
+  *     that results in the most pure pair of sub-groups.  The criteria for evaluating
+  *     splits is different for the different forest types.
+  *   - Choose the best split, and create a split-node (branch) based on that split.
+  *   - Create a left and right child node at the next level under the split-node.
+  *   - Reallocate the data points from the node to the left or right child nodes.
+  *   - If any group is pure (i.e. same dependent value for all members) or we've reached
+  *     maxDepth, replace all the datapoints in that node with a single leaf node.
+  * - At the end of the process, all the datapoints will have been absorbed by leaf nodes,
+  *   and the resulting trees will contain only split-nodes and leaf-nodes.
+  * - Note that a single common data structure (see LT_Types.TreeNodeData) is used to
+  *   represent:
+  *   - Data-points that need to be processed for a node (only during tree growing)
+  *   - A split-node
+  *   - A leaf-node
+  * - The resulting tree-nodes representing the forest are converted to a Forest Model
+  *   for later use in prediction or analysis.
   */
 EXPORT RF_Base(DATASET(GenField) X_in=DATASET([], GenField),
               DATASET(GenField) Y_In=DATASET([], GenField),
               UNSIGNED numTrees=100,
               UNSIGNED featuresPerNodeIn=0,
               UNSIGNED maxDepth=255) := MODULE
+  // Resequence ids to go from 1-numRecords.
+  SHARED resequenceByWi(DATASET(GenField) dat) := FUNCTION
+    // Dat should be evenly distributed at this point
+    xGen := RECORD(GenField)
+      t_RecordId newId := 0;
+    END;
+    xDat := PROJECT(dat, TRANSFORM(xGen, SELF := LEFT));
+    xGen setNewIds(xGen l, xGen r) := TRANSFORM
+      newId := IF(l.wi != r.wi, 1, IF(r.id = l.id, l.newId, l.newId + 1));
+      SELF.newId := newId;
+      SELF := r;
+    END;
+    xDatS := SORT(xDat, wi, id);
+    xDat2 := ITERATE(xDatS, setNewIds(LEFT, RIGHT));
+    outDat0 := PROJECT(xDat2, TRANSFORM(GenField, SELF.id := LEFT.newId, SELF := LEFT));
+    outDat := DISTRIBUTE(outDat0, HASH32(wi, id));
+    return outDat;
```

Comment:
This function will perform a global iterate.  As written, the nodes will process the data in serial instead of parallel because the value for newID on the k-th record cannot be established without the value from the (k-1)-th record on a global basis.  If the work-items cannot get too large, then this is OK.  If work-items can be large enough for this to be a performance problem, then consider two passes.

Reply:
I could only think of two ways to assign sequential numbers:
GROUP -> PROJECT or the global ITERATE.  It seems like the ITERATE would scale better because of the child record limitations on GROUP.  Is there another way to handle this?  This is iterating over the independent data, so if there's a size limitation, that would limit the number of records that could be supported.

Reply:
There are several more, though they are variations of each other.  You are correct on GROUP as it would be useful only when the number of work items were large and there were a modest number of records per group.

The basic idea is to enumerate the records in parallel on each node from 1 to the number of records on the node.  You can use a local table, or a CHOOSEN(..., LOCAL), or a DEDUP(..., RIGHT, LOCAL) to get the node level counts.  A record set with one record per node can always fit on a single node, so you can do a local iterate to create adjustment numbers.  This small dataset of adjustments can then be joined with the large dataset in a lookup to perform the final adjustment.

An alternative to an explicit DISTRIBUTE(..., 0) would be for each node to compute the adjustment table in parallel by doing a self join using ALL or LOOKUP, MANY to make the platform perform the replication and distribution.  Jake has been making improvements to this basic mechanism to reduce latency.

Your case is a little more complicated because you want the numbers to restart at each work item.  So, you will have one of more adjustment records per node.  I can find some examples next week if you would like to see the approach in action.

Reply:
An update.  There is a function macro in ML_Core to enumerate within some field groupers.  Utils.SequenceInFields(...)